hazelcast/src/main/java/com/hazelcast/sql/impl/expression/CaseExpression.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.predicate.TernaryLogic;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+
+public class CaseExpression<T> implements Expression<T>, IdentifiedDataSerializable {
+    private Expression<Boolean>[] whenExpressions;
+    private Expression<?>[] thenExpressions;
+    private Expression<?> elseExpression;
+    private QueryDataType resultType;
```

Comment:
Do we really need the `resultType` here? Couldn't it be inferred from one of the return types?